Fix preview of theming

danxuliu opened the pull request:
Pull request #5429 made cached SCSS files depend on a hash of the base URL, so the _/css/core/server.css_ file does no longer exist; as the file can not be loaded the _"Loading preview"_ message is never removed and the _"Saved"_ message is never shown.

As it now depends on the hash of the base URL the file to be reloaded can no longer be hardcoded, so the DataResponse from the controller now provides the full URL to the _server.css_ file that has to be reloaded (if any).

Fixes #5975

Acceptance tests for the preview after setting and resetting the color in the Theming app are also included in this pull request.

danxuliu committed:
Add getter for the timeout multiplier

In some cases the acceptance tests have to explicitly wait for something
to happen without using the "find" method from the actor; in those cases
the timeout multiplier needs to be taken into account too, so the test
cases must be able to retrieve it from the actor.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add acceptance tests for setting the color in the Theming app

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Refactor reloadStylesheets to receive absolute URLs

This is a preparatory step for a following change in which
reloadStylesheets will have to be able to receive absolute URLs.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Fix preview of theming

Pull request #5429 made cached SCSS files depend on a hash of the base
URL, so the "/css/core/server.css" file does no longer exist; as the
file can not be loaded the "Loading preview" message is never removed
and the "Saved" message is never shown.

As it now depends on the hash of the base URL the file to be reloaded
can no longer be hardcoded, so the full URL to the "server.css" file
that has to be reloaded (if any) is now got from the DataResponse
provided by the controller.

Fixes #5975

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu commented:
Oops, I had forgotten to add the acceptance tests for the Theming app to Drone; it is fixed now.

MorrisJobke commented:
You are awesome!!!! 🚀 😍

Thanks again for helping that much 👍

MorrisJobke commented on apps/theming/js/settings-admin.js:
```diff
@@ -62,8 +62,18 @@ function preview(setting, value) {
+		// Ensure that the URL starts by '/' to prevent "stylesheet.load" from
+		// prepending the URL of the current page.
+		if (serverCssUrl[0] !== '/') {
+			serverCssUrl = '/' + serverCssUrl;
```
This doesn't work for instances that are in a subfolder. It needs to be the full URL. linkToAbsolute or something like that should exist, which prepends you the webroot.

danxuliu replied:
> This doesn't work for instances that are in a subfolder. It needs to be the full URL.

Totally right; [getCachedSCSS removes the webroot](https://github.com/nextcloud/server/blob/517dfdda0dfe928836e32db2dfffe04545abfe8e/lib/private/Template/SCSSCacher.php#L297).

> linkToAbsolute or something like that should exist, which prepends you the webroot.

I have used `Util::linkTo` instead, as `Util::linkToAbsolute` adds the webroot, but also the protocol and server host.

MorrisJobke reviewed (approved):
Tested and works 👍

nickvergessen commented:
Tests fail:
```
There were 8 failures:

1) OCA\Theming\Tests\Controller\ThemingControllerTest::testUpdateStylesheet
```

nickvergessen reviewed (dismissed):
Tests

MorrisJobke commented:
And please rebase on master - we updated the drone config 🙈

danxuliu committed:
Split updateStylesheet tests for success and error

This is a preparatory step for a following commit in which the
properties present in the response will change depending on whether the
request was successful or not.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Provide "server.css" URL in ThemingController response

Pull request #5429 made cached SCSS files depend on a hash of the base
URL, so the "/css/core/server.css" file does no longer exist. The
"server.css" URL must be known by the Theming app in order to update the
stylesheets when previewing the changes to the theme, so the
DataResponse from the controller now provides the full URL to the
"server.css" file that has to be reloaded (if any).

The "server.css" URL provided by the response will be taken into account
by the JavaScript front-end in a following commit.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Unify DataResponse format with its uses in the rest of the file

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu commented:
@nickvergessen 
> Tests fail

:facepalm: Yes, I forgot to check the tests. And not only that, I also used a deprecated method (`OC\Util::linkTo`). Both are fixed now.

@MorrisJobke 
> And please rebase on master - we updated the drone config 🙈

Done.

MorrisJobke commented:
> Edit: By the way, I have noticed that the commit order in the GitHub interface is messed up (as it is ordered chronologically instead of parent->child); I can change their timestamp and push again so they are showed in the right order for easier review if you want.

No - just leave it as it is :) That's fine :)

MorrisJobke reviewed (approved):
Tested again and works 👍 Code looks good

MorrisJobke commented:
Huge thanks again @danxuliu for this awesome help here 👍 

Could I ask you to open  the backport to stable12? That would be nice :)

danxuliu commented:
> Huge thanks again @danxuliu for this awesome help here 👍

My pleasure :-)

> Could I ask you to open the backport to stable12? That would be nice :)

Here you have (yes, I had it prepared already :-P ): #6067
